control-plane/pkg/reconciler/sink/kafka_sink.go

```diff
@@ -0,0 +1,162 @@
+ * limitations under the License.
+ */
+
+package sink
+
+import (
+	"context"
+	"math"
+
+	"github.com/Shopify/sarama"
+	"go.uber.org/zap"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/util/retry"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/reconciler"
+
+	eventing "knative.dev/eventing-kafka-broker/control-plane/pkg/apis/eventing/v1alpha1"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/config"
+	coreconfig "knative.dev/eventing-kafka-broker/control-plane/pkg/core/config"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/log"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/receiver"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/base"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/kafka"
+)
+
+type Reconciler struct {
+	*base.Reconciler
+
+	ConfigMapLister corelisters.ConfigMapLister
+
+	// ClusterAdmin creates new sarama ClusterAdmin. It's convenient to add this as Reconciler field so that we can
+	// mock the function used during the reconciliation loop.
+	ClusterAdmin kafka.NewClusterAdminFunc
+
+	Configs *config.Env
+}
+
+func (r *Reconciler) ReconcileKind(ctx context.Context, ks *eventing.KafkaSink) reconciler.Event {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		return r.reconcileKind(ctx, ks)
+	})
+}
+
+func (r *Reconciler) reconcileKind(ctx context.Context, ks *eventing.KafkaSink) error {
+
+	logger := log.Logger(ctx, "reconcile", ks)
+
+	statusConditionManager := base.StatusConditionManager{
+		Object:     ks,
+		SetAddress: ks.Status.SetAddress,
+		Configs:    r.Configs,
+		Recorder:   controller.GetEventRecorder(ctx),
+	}
+
+	if ks.Spec.NumPartitions != nil && ks.Spec.ReplicationFactor != nil {
+		topicConfig := &kafka.TopicConfig{
+			TopicDetail: sarama.TopicDetail{
+				NumPartitions:     *ks.Spec.NumPartitions,
+				ReplicationFactor: *ks.Spec.ReplicationFactor,
+			},
+			BootstrapServers: kafka.BootstrapServersArray(ks.Spec.BootstrapServers),
+		}
+
+		topic, err := r.ClusterAdmin.CreateTopic(logger, ks.Spec.Topic, topicConfig)
+		if err != nil {
+			return statusConditionManager.FailedToCreateTopic(topic, err)
+		}
+	}
+	statusConditionManager.TopicCreated(ks.Spec.Topic)
+
+	logger.Debug("Topic created", zap.Any("topic", ks.Spec.Topic))
+
+	// Get sinks config map.
+	sinksConfigMap, err := r.GetOrCreateDataPlaneConfigMap()
+	if err != nil {
+		return statusConditionManager.FailedToGetConfigMap(err)
+	}
+
+	logger.Debug("Got sinks config map")
+
+	// Get sinks data.
+	sinks, err := r.GetDataPlaneConfigMapData(logger, sinksConfigMap)
+	if err != nil && sinks == nil {
+		return statusConditionManager.FailedToGetDataFromConfigMap(err)
+	}
+	if sinks == nil {
+		sinks = &coreconfig.Sinks{}
+	}
+
+	logger.Debug(
+		"Got sinks data from config map",
+		zap.Any("sinks", log.NewSinksMarshaller(sinks)),
+	)
+
+	// Get sink configuration.
+	// TODO(pierDipi) support content modes
```

Comment:
DYM on the data plane contract? We can do it in a future pr, maybe after fixing #142

Reply:
This is done already!